firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java

```diff
@@ -14,32 +14,257 @@
+
+    // Headers required for Android API Key Restrictions.
+    httpURLConnection.setRequestProperty(X_ANDROID_PACKAGE_HEADER, context.getPackageName());
+    httpURLConnection.setRequestProperty(X_ANDROID_CERT_HEADER, getFingerprintHashForPackage());
+
+    // Header to denote request is retryable on the server.
+    httpURLConnection.setRequestProperty(X_GOOGLE_GFE_CAN_RETRY, "yes");
+
+    // Header to tell server that client expects stream response
+    httpURLConnection.setRequestProperty(X_ACCEPT_RESPONSE_STREAMING, "true");
+
+    // Headers to denote that the request body is a JSONObject.
+    httpURLConnection.setRequestProperty("Content-Type", "application/json");
+    httpURLConnection.setRequestProperty("Accept", "application/json");
+  }
+
+  private JSONObject createRequestBody() {
+    Map<String, String> body = new HashMap<>();
+    body.put(
+        "project", extractProjectNumberFromAppId(this.firebaseApp.getOptions().getApplicationId()));
+    body.put("namespace", this.namespace);
+    body.put(
+        "lastKnownVersionNumber", Long.toString(configFetchHandler.getTemplateVersionNumber()));
+    return new JSONObject(body);
+  }
+
+  private void setRequestParams(HttpURLConnection httpURLConnection) throws IOException {
+    httpURLConnection.setRequestMethod("POST");
+    byte[] body = createRequestBody().toString().getBytes("utf-8");
+    OutputStream outputStream = new BufferedOutputStream(httpURLConnection.getOutputStream());
+    outputStream.write(body);
+    outputStream.flush();
+    outputStream.close();
+  }
+
+  private synchronized boolean canMakeHttpStreamConnection() {
+    return !listeners.isEmpty() && !isStreamOpen;
+  }
+
+  private String getRealtimeURL(String namespace) {
+    return String.format(
+        REALTIME_REGEX_URL,
+        extractProjectNumberFromAppId(firebaseApp.getOptions().getApplicationId()),
+        namespace);
+  }
+
+  private URL getUrl() {
+    URL realtimeURL = null;
+    try {
+      realtimeURL = new URL(getRealtimeURL(namespace));
+    } catch (MalformedURLException ex) {
+      Log.i(TAG, "URL is malformed");
+    }
+
+    return realtimeURL;
+  }
+
+  private synchronized HttpURLConnection makeHttpConnection() {
+    URL realtimeUrl = getUrl();
+    try {
+      this.httpURLConnection = (HttpURLConnection) realtimeUrl.openConnection();
+      setCommonRequestHeaders(httpURLConnection);
+      setRequestParams(httpURLConnection);
+    } catch (IOException ex) {
+      for (ConfigUpdateListener listener : listeners) {
+        listener.onError(
+            new FirebaseRemoteConfigRealtimeUpdateStreamException(
+                "Can't establish http stream.", ex.getCause()));
+      }
+    }
+
+    return httpURLConnection;
+  }
+
+  private synchronized void startAutoFetch() {
+    ConfigUpdateListener retryCallback =
+        new ConfigUpdateListener() {
+          @Override
+          public void onEvent() {
+            pauseRealtime();
```

Comment:
In my next PR I'll introduce the http retry method. So when autofetch and the http stream connection close we want to explicitly close the http connection(.disconnect) using pauseRealtime so that we can start a new one with the http retry method.